### ISSUE TITLE
attack on row contribution outliers

### ISSUE THREAD

**Issue by yoid2000** [diffix/test-syndiffix#5]
@pdobacz  before you start digging into `anonymeter`, I think it would be good to run the attack that Cristian mentioned in the meeting today. 

You can find a description of the attack in section 5.20, page 35, of https://arxiv.org/pdf/2201.04351.pdf.

What you'll have to do is create a bunch of datasets with two columns each. One column is categorical with a relatively small number of distinct values (you can test 2, 5, and 10 values). The other column is the AID column. It has multiple rows per entity, but all rows for a given entity have the same categorical value. Select each AID's category randomly from the set of categories. (This selection can also be skewed, but let's not worry about that now.)

All entities are assigned some number of rows taken from a uniform distribution of some range (say between 50 and 100 rows). One outlier entity, however, is assigned substantially more rows, like 200 or even more.

Actually, you'll need to build two datasets for each test. One dataset is as described above. The other has one row per entity (and the same category assignment).

Build synthetic datasets from both tables. From the one-row-per-AID table, you estimate the number of distinct AIDs per category, and the total number of distinct AIDs. From the multi-row table, you determine to total number of rows.

You then estimate the expect number of rows per category on the assumption that all entities have the same number of rows. You then select the category that exceeds that estimate by the largest amount. (If the categories are uniformly assigned to AIDs, then this is simply the category with the most rows.)

You then guess that this category is the one assigned to the outlier AID.

Do this for a few hundred different datasets, and record the fraction of times the guess is right.

**Comment by pdobacz** [comment]
A bunch of clarifications needed:
1. We synthesize the AID column, correct? Is this in line with all the assumptions and also - how do we do this with Syndiffix? just include it in the synthesized columns instead of setting this to be the AID column?
2. Why is it natural to expect that both synthesized data sets will be made available to the attacker like this?
3. Related - is inferring "number of distinct AIDs per category, and the total number of distinct AIDs" from the multi-row synthetic table too inaccurate for the attack to work?
4. I'm reading the section from the paper (specifically the "Attack" subsection) and knowing somewhat how this should go I find it hard to parse. Is there some word missing there, like it should read "multiplying the number of protected entities **[per bucket]** by the average number of rows per protected entity"?
5. Do you think there's no point in trying to reuse the attack code cited in the paper? (possibly the answers to the above will answer this one, but asking nevertheless)

**Comment by yoid2000** [comment]
> Why is it natural to expect that both synthesized data sets will be made available to the attacker like this?

A natural thing would be two tables, one listing user attributes (birth year, gender, title, year hired, etc. etc.), and another time-series one that can be JOINed to it (mobile data locations, purchases, or whatever). It would be very natural to synthesize each of these tables individually as well as joined.

(More generally, we consider all kinds of attacks, even those where the information needed by the attacker is unlikely though not impossible to exist.)

**Comment by yoid2000** [comment]
> I'm reading the section from the paper (specifically the "Attack" subsection) and knowing somewhat how this should go I find it hard to parse. Is there some word missing there, like it should read "multiplying the number of protected entities [per bucket] by the average number of rows per protected entity"?

Yes, you are correct.

**Comment by yoid2000** [comment]
> Do you think there's no point in trying to reuse the attack code cited in the paper? (possibly the answers to the above will answer this one, but asking nevertheless)

The GDA-score code assumes a Diffix interface, so you'd have to first update the GDA-score code to deal with synthetic data, and then modify the attack with that underlying assumption.

So the short answer is almost certainly 'no'.

**Comment by pdobacz** [comment]
> We synthesize the AID column, correct? Is this in line with all the assumptions and also - how do we do this with Syndiffix? just include it in the synthesized columns instead of setting this to be the AID column?

@yoid2000 What about this one? I feel like I'm missing something here. Either we're dropping the information about AID we need to proceed with the attack or we use Syndiffix (and any other method, actually) a little bit non-standard by synthesizing the AID column instead of letting it be the AID.

**Comment by pdobacz** [comment]
this is my first result, but I need to sleep on it and review this once more to not be doing anything stupid. It's also excruciatingly slow, for a simple thing like this, I need to see if it can be made more conveniently fast to run.

```
   nCategories  guessCorrect
0            2          0.70
1            5          0.49
2           10          0.21
```

It decays as nCategories increases, but I was expecting it to decay faster, given that we should be flattening the outlier :thinking:

**Comment by yoid2000** [comment]
I don't expect flattening to fully make the victim undetectable, though I
might have expected it to work a little better than this. So long as the
probability is relatively small, then our solution is good enough. A 0.7
success rate on a 0.5 baseline probability is only 40% increase. In
general, I regard anything below 50% increase as acceptable.

Still we might want to dig in a bit and make sure outlier flattening is
indeed working correctly.

Don't worry particularly about making it fast. We won't run this often.

I'm more interested in how the ctgan and mostly.ai do...

PF

**Comment by pdobacz** [comment]
OK, I validated the implementation a bit, it seemed fine for a bunch of sanity checks, PR coming soon. Here are the results (I needed to process only 100 datasets, and also reduce the row-per-aid factors by factor of 10 (down to 5-10 rows per regular AID, 20 for the outlier), because TVAE takes its time :/. I can run a bigger setup overnight or on cluster if needed.

`original` as method is just doing the guesses from original data, one of the sanity checks.
`elapsedSeconds` is mean time for synthesizing one dataset (i.e. a pair of tables).

```
   nCategories     method  guessCorrect  elapsedSeconds
0            2   original          0.84        0.000128
1            2  syndiffix          0.68        1.257005
2            2       tvae          0.42        8.178638
3            5   original          0.80        0.000131
4            5  syndiffix          0.53        1.219762
5            5       tvae          0.30        8.054284
6           10   original          0.92        0.000137
7           10  syndiffix          0.36        1.364611
8           10       tvae          0.23        9.061641
```

I now noticed you requested ctgan and mostly.ai not tvae, that one stuck to my head :facepalm: 

I'll run for ctgan and see about mostly.ai too.

**Comment by pdobacz** [comment]
Ah, I spotted one error for TVAE - I'm including an auxiliary column on top of the category, these results might be off, hold on, lemme fix this

**Comment by pdobacz** [comment]
@yoid2000 

I have rerun both the local and global outliers (local using #11 changes) using 1000 datasets. I'll put down the SLURM scripts and commands which led me to this result, as well as the results:

### Global outlier

nAids: 100, nDatasets: 1000, outlierCount: 200

```
    nCategories     method  guessCorrect  elapsedSeconds
0             2      ctGan         0.763          83.432
1             2   original         1.000           0.000
2             2  syndiffix         0.570          10.533
3             2       tvae         0.998          28.927
4             5      ctGan         0.658          55.907
5             5   original         1.000           0.000
6             5  syndiffix         0.329          10.554
7             5       tvae         0.820          15.237
8            10      ctGan         0.728          51.310
9            10   original         1.000           0.000
10           10  syndiffix         0.190           9.503
11           10       tvae         0.832          14.958
```

### Local outlier

nAids: 200, nDatasets: 999, because one failed tvae, outlierCount 200

```
    nCategories     method  guessCorrect  elapsedSeconds
0             4      ctGan         0.826         426.146
1             4   original         1.000           0.000
2             4  syndiffix         0.599           6.121
3             4       tvae         0.909         418.415
4            10      ctGan         0.564         439.680
5            10   original         1.000           0.000
6            10  syndiffix         0.357           6.104
7            10       tvae         0.372         384.954
8            20      ctGan         0.043         440.840
9            20   original         1.000           0.000
10           20  syndiffix         0.207           6.170
11           20       tvae         0.601         389.209
```


### SLURM script:

Global outlier:
```
#!/usr/bin/bash
#SBATCH --time=600
#SBATCH --chdir=<path to test_syndiffix source dir>
#SBATCH --array=0-999      
#SBATCH --cpus-per-task=4

arrayNum="${SLURM_ARRAY_TASK_ID}"
syndiffixPath=<path to SynDiffix F# project>

source ~/venvtest-syndiffix/bin/activate

/usr/bin/time -v \
  python \
  attacks/outlierInference.py \
  one $syndiffixPath \
  [2,5,10] 100 $arrayNum 200 False
```

Local same as global, only with `[4,10,20] 200 $arrayNum 200 True` in the last line to configure `outlierInference.py`.

Slurm scripts both produce results in the `data` subdir of the working directory (take care to not overwrite). Afterwards `python  attacks/outlierInference.py  collect --nDatasets=1000` produces the output.

**Comment by yoid2000** [comment]
There is a ctGan value that doesn't make sense:

```
8            20      ctGan         0.043         440.840
```

Any idea why this might be?

PF

**Comment by pdobacz** [comment]
Yeah, I've seen it too, but no idea. CTGAN looked very creative in the taxi case, maybe it decides to remove the outlier aggressively here. Should we have an issue to investigate this?

**Comment by pdobacz** [comment]
> Could you post these results in the repo? (in the readme or something) Thanks, PF

I would rather not. If we post them there, then we either need to keep them up-to-date or they become stale (and there's a bunch of stale content in the readme already). Here they're not expected to be updated.

**Comment by yoid2000** [comment]
You mean "here" in the issue itself? My main concern is that if I want to look these up again later I won't remember where to look. Perhaps we could put a pointer to the issue in the readme???

**Comment by pdobacz** [comment]
I suggest a pointer in the attack script, it should be natural to find, but pointer in readme is also an alternative.

**Comment by yoid2000** [comment]
yes, that is fine. thanks.
